src/StatusWriter.cxx

```diff
@@ -0,0 +1,144 @@
+  std::cout << "stream_master :\t" << data.status << "\n";
+  for (size_t i = 0; i < data.topic.size(); ++i) {
+    std::cout << data.topic[i] << ":\n";
+    print(data.streamer_status[i]);
+    print(data.streamer_stats[i]);
+    std::cout << "\n";
+  }
+}
+
+void StdIOWriter::print(const StreamerStatusType &x) {
+  std::cout << "\tstatus :\n"
+            << "\t\tmessages :\t" << x.messages << "\n"
+            << "\t\tbytes :\t\t" << x.bytes << "\n"
+            << "\t\terrors :\t" << x.errors << "\n";
+}
+void StdIOWriter::print(const StreamerStatisticsType &x) {
+  std::cout << "\tstatistics :\n"
+            << "\t\tsize average :\t" << x.size_avg << "\n"
+            << "\t\tsize std :\t" << x.size_std << "\n"
+            << "\t\tfrequency average :\t" << x.freq_avg << "\n"
+            << "\t\tfrequency std :\t\t" << x.freq_std << "\n";
+}
+
+rapidjson::Document JSONWriter::write(const StreamMasterStatus &data) {
+  return write_impl(data);
+}
+
+rapidjson::Document JSONWriter::write_impl(const StreamMasterStatus &data) {
+  using namespace rapidjson;
+  Document d;
+  auto &a = d.GetAllocator();
+  d.SetObject();
+  { // message type
+    d.AddMember("type", "filewriter_streammaster_status", a);
+  }
+  { // stream master info
+    Value sm,ss;
+    sm.SetObject();
+    std::string s{Err2Str(StreamMasterError{data.status})};
+    ss.SetString(s.c_str(),s.size(),a);
+    sm.AddMember("status", ss, a);
+    d.AddMember("streammaster", sm, a);
+  }
+  { // streamers info
+    Value ss;
+    ss.SetObject();
+    for (size_t i = 0; i < data.topic.size(); ++i) {
+      Value key{data.topic[i].c_str(), a};
+      Value val;
+      val.SetObject();
+      val.AddMember("status", to_json(data.streamer_status[i], d), a);
+      val.AddMember("statistics", to_json(data.streamer_stats[i], d), a);
+      ss.AddMember(key, val, a);
+    }
+    d.AddMember("streamer", ss, a);
+  }
+  return d;
+}
+
+rapidjson::Value JSONWriter::to_json(const StreamerStatusType &x,
+                                     return_type &d) {
+  using namespace rapidjson;
+  auto &a = d.GetAllocator();
+  Value value;
+  value.SetObject();
+  value.AddMember("messages", int(x.messages), a);
+  value.AddMember("bytes", int(x.bytes), a);
+  value.AddMember("errors", int(x.errors), a);
+  return value;
+}
+
+rapidjson::Value JSONWriter::to_json(const StreamerStatisticsType &x,
+                                     return_type &d) {
+  using namespace rapidjson;
+  auto &a = d.GetAllocator();
+
+  Value value;
+  value.SetObject();
+  {
+    Value size;
+    size.SetObject();
+    size.AddMember("avg", x.size_avg, a);
+    size.AddMember("std", x.size_std, a);
+    value.AddMember("size", size, a);
+  }
+  {
+    Value freq;
+    freq.SetObject();
+    freq.AddMember("avg", x.freq_avg, a);
+    freq.AddMember("std", x.freq_std, a);
+    value.AddMember("freq", freq, a);
+  }
+  return value;
+}
+
+std::string JSONStreamWriter::write(const StreamMasterStatus &data) {
```

Comment:
Or, if code sharing is the goal, then do not share code via inheritance but via composition.